### PR TITLE
End to end example with tf transform with Dataflow and CMLE

### PR DESCRIPTION
**Pull request by lfloretta**
Hi @puneith and @thedriftofwords,

here is again an end to end example based on census data that shows how to perform feature engineering with tf transform with DirectRunner or DataflowRunner.

The tensorflow model is the same of census/estimators but with different input function and serving function as both are indeed build using tf transform. The serving function has the same signaturedef of census/estimators for the json format.

### CODE REVIEW

File: census/tftransformestimator/trainer/model.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    """Help function to parse tf.Example"""
+    parsed = tf.parse_single_example(record, transformed_feature_spec)
+    label = parsed.pop(LABEL_KEY)
+    return parsed, label
+
+  def input_fn():
+    """Input function for training and eval."""
+    #See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/docs_src/performance/datasets_performance.md
+    files = tf.data.Dataset.list_files(os.path.join(tft_working_dir, filebase + '*'))
+    dataset = files.interleave(lambda x : tf.data.TFRecordDataset(x).map(parser,
```
Comment (davidcavazos) [inline census/tftransformestimator/trainer/model.py]: Can you please split this into a single operation per line?
```python
files = ...
dataset = files.interleave(tf.data.TFRecordDataset)
dataset = dataset.map(parser)
```

File: census/tftransformestimator/trainer/model.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    buffer_size: Buffer size for the shuffle
+    prefetch_buffer_size: Number of example to prefetch
+
+  Returns:
+    The input function for training or eval.
+  """
+  if buffer_size is None:
+    buffer_size = 2 * batch_size + 1
+
+  # Examples have already by transformed so we only need the feature_columns
```
Comment (dizcology) [inline census/tftransformestimator/trainer/model.py]: already "been" transformed.

File: census/tftransformestimator/trainer/model.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    raw_placeholder_spec = RAW_DATA_METADATA.schema.as_batched_placeholders()
+    # remove label key that is not going to be available at seving
+    raw_placeholder_spec.pop(LABEL_KEY)
+
+    raw_input_fn = tf.estimator.export.build_raw_serving_input_receiver_fn(
+        raw_placeholder_spec)
+    raw_features, receiver_tensors , _ = raw_input_fn()
+
+    _, transformed_features = (
+        saved_transform_io.partially_apply_saved_transform(
```
Comment (dizcology) [inline census/tftransformestimator/trainer/model.py]: This seems to be the counterpart of your comments on preprocessing.py line 172-174, where transform_fn was written to GCS.  Please add some comments here to highlight that.

File: census/tftransformestimator/README.md [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+                                    --train-filebase train \
+                                    --eval-filebase test \
+                                    --train-steps $TRAIN_STEPS \
+                                    --job-dir $GCS_JOB_DIR
+
+```
+
+# Hyperparameter Tuning
+Cloud ML Engine allows you to perform Hyperparameter tuning to find out the
+most optimal hyperparameters. See [Overview of Hyperparameter Tuning]
```
[the remark this answers is not in the document]
Comment (lfloretta) [reply census/tftransformestimator/README.md]: Solved

File: census/tftransformestimator/preprocess.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        an int64 Tensor representing the label index for classification tasks
+      """
+      table = lookup.index_table_from_tensor(['<=50K', '>50K'])
+      return table.lookup(label)
+
+    outputs[LABEL_KEY] = tft.apply_function(convert_label, inputs[LABEL_KEY])
+    return outputs
+
+  def fix_comma_and_filter_third_column(line):
+    cols = line.replace(', ', ',').split(',')
```
[the remark this answers is not in the document]
Comment (lfloretta) [reply census/tftransformestimator/preprocess.py]: Solved

File: census/tftransformestimator/preprocess.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+          coder=example_proto_coder.ExampleProtoCoder(
+              transformed_metadata.schema))
+
+      # Will write a SavedModel and metadata to two subdirectories of
+      # working_dir, given by transform_fn_io.TRANSFORM_FN_DIR and
+      # transform_fn_io.TRANSFORMED_METADATA_DIR respectively.
+      _ = (
+          transform_fn
+          | 'WriteTransformFn' >>
+          transform_fn_io.WriteTransformFn(working_dir))
```
[the remark this answers is not in the document]
Comment (lfloretta) [reply census/tftransformestimator/preprocess.py]: Solved

File: census/tftransformestimator/preprocess.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+        args.project is None or
+        args.region is None):
+      parser.error(('--runner DataflowRunner requires --job-name, '
+                    '--project, --region, --staging-location',
+                    '--temp-location')
+      )
+
+
+
+    options={
```
[the remark this answers is not in the document]
Comment (lfloretta) [reply census/tftransformestimator/preprocess.py]: I implemented the change

File: census/tftransformestimator/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,46 @@
+apache-beam==2.2.0
```
Comment (davidcavazos) [inline census/tftransformestimator/requirements.txt]: Can we test if this works with the latest versions of all libraries?

File: census/tftransformestimator/trainer/task.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
+import argparse
+import os
+
+#import trainer.model as model
```
Comment (davidcavazos) [inline census/tftransformestimator/trainer/task.py]: Remove if not needed?

File: census/tftransformestimator/trainer/task.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+import model
+
+import tensorflow as tf
+from tensorflow.contrib.training.python.training import hparam
+
+
+def run_experiment(hparams):
+  """Run the training and evaluate using the high level API"""
+  print("run")
```
Comment (davidcavazos) [inline census/tftransformestimator/trainer/task.py]: Do we want this here? Maybe we could provide more some useful information, or remove it.

Comment (lfloretta) [reply census/tftransformestimator/trainer/task.py]: Done

File: census/tftransformestimator/trainer/task.py (hunk 9 above)
Comment (lfloretta) [reply census/tftransformestimator/trainer/task.py]: Done

File: census/tftransformestimator/requirements.txt (hunk 8 above)
Comment (lfloretta) [reply census/tftransformestimator/requirements.txt]: Hi David, 
moving to the new versions of all libraries require minor changes to requirement.txt and requirements_dataflow.txt. However, I would suggest to postpone them as the rest of the examples are still in tf 1.4.

Speak soon,
Lucio

File: census/tftransformestimator/README.md [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+                             --train-steps $TRAIN_STEPS \
+                             --job-dir $OUTPUT_DIR
+
+```
+
+### Using Cloud ML Engine
+Run the code on Cloud ML Engine using `gcloud`. Note how `--job-dir` comes
+before `--` while training on the cloud and this is so that we can have
+different trial runs during Hyperparameter tuning.
+
```
Comment (thedriftofwords) [inline census/tftransformestimator/README.md]: Consider adding link to a further explanation of these training flags in the docs:

"See more information about [training job arguments](https://cloud.google.com/ml-engine/docs/tensorflow/training-jobs#submitting_the_job)."

File: census/tftransformestimator/README.md [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+```
+tensorboard --logdir=$GCS_JOB_DIR
+```
+
+## Accuracy and Output
+You should see the output for default number of training steps and approx accuracy close to `80%`.
+
+# Distributed Node Training
+Distributed node training uses [Distributed TensorFlow](https://www.tensorflow.org/deploy/distributed).
+The main change to make the distributed version work is usage of [TF_CONFIG](https://cloud.google.com/ml/reference/configuration-data-structures#tf_config_environment_variable)
```
Comment (thedriftofwords) [inline census/tftransformestimator/README.md]: Please update the docs link to TF_CONFIG: https://cloud.google.com/ml-engine/docs/tensorflow/distributed-training-details

File: census/tftransformestimator/README.md [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+environment variable. The environment variable is generated using `gcloud` and parsed to create a
+[ClusterSpec](https://www.tensorflow.org/deploy/distributed#create_a_tftrainclusterspec_to_describe_the_cluster). See the [ScaleTier](https://cloud.google.com/ml/pricing#ml_training_units_by_scale_tier) for predefined tiers
+
```
Comment (thedriftofwords) [inline census/tftransformestimator/README.md]: Please update the docs link to ScaleTier: https://cloud.google.com/ml-engine/docs/pricing#scale-tier

File: census/tftransformestimator/README.md [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+                                    --tft-working-dir $TFT_OUTPUT_DIR \
+                                    --train-filebase train \
+                                    --eval-filebase test \
+                                    --train-steps $TRAIN_STEPS
+
+```
+
+# Hyperparameter Tuning
+Cloud ML Engine allows you to perform Hyperparameter tuning to find out the
+most optimal hyperparameters. See [Overview of Hyperparameter Tuning](https://cloud.google.com/ml/docs/concepts/hyperparameter-tuning-overview) for more details.
```
Comment (thedriftofwords) [inline census/tftransformestimator/README.md]: Please update the hyperparameter tuning docs link: https://cloud.google.com/ml-engine/docs/tensorflow/hyperparameter-tuning-overview

File: census/tftransformestimator/README.md [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+
+```
+export TRAIN_STEPS=1000
+DATE=`date '+%Y%m%d_%H%M%S'`
+export OUTPUT_DIR=census_$DATE
+rm -rf $OUTPUT_DIR
+```
+
+### Model location reuse
+Its worth calling out that unless you want to reuse the old model output dir,
```
[the remark this answers is not in the document]
Comment (lfloretta) [reply census/tftransformestimator/README.md]: I rephrased as "Its worth calling out that unless you want to restore training from the
[checkpoints](https://www.tensorflow.org/get_started/checkpoints) saved 
in the old model output dir, model location should be a new location."

File: census/tftransformestimator/README.md [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+
+### Disclaimer
+This dataset is provided by a third party. Google provides no representation,
+warranty, or other guarantees about the validity or any other aspects of this dataset.
+
+### Set Environment Variables
+Please run the export and copy statements first:
+
+```
+TRAIN_FILE=gs://cloudml-public/census/data/adult.data.csv
```
[the remark this answers is not in the document]
Comment (lfloretta) [reply census/tftransformestimator/README.md]: Done

File: census/tftransformestimator/README.md (hunk 11 above)
Comment (lfloretta) [reply census/tftransformestimator/README.md]: Done

File: census/tftransformestimator/README.md (hunk 12 above)
Comment (lfloretta) [reply census/tftransformestimator/README.md]: Done

File: census/tftransformestimator/README.md (hunk 13 above)
Comment (lfloretta) [reply census/tftransformestimator/README.md]: Done

File: census/tftransformestimator/README.md (hunk 14 above)
Comment (lfloretta) [reply census/tftransformestimator/README.md]: Done

File: census/tftransformestimator/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (thedriftofwords) [reply census/tftransformestimator/README.md]: SGTM thanks
